File: generic/ReLU6.c
#ifndef TH_GENERIC_FILE
#define TH_GENERIC_FILE "generic/ReLU6.c"
#else

void THNN_(ReLU6_updateOutput)(
          THNNState *state,
          THTensor *input,
          THTensor *output,
          bool inplace)
{
  if (inplace)
  {
    TH_TENSOR_APPLY(real, input,
      if (*input_data <= 0)
        *input_data = 0;
      else if (*input_data >= 6)
        *input_data = 6;
    );
    THTensor_(set)(output, input);
  }
  else
  {
    THTensor_(resizeAs)(output, input);
    TH_TENSOR_APPLY2(real, output, real, input,
      *output_data =
         (*input_data > 0) ? ((*input_data < 6) ? *input_data : 6) : 0;
    );
  }
}

void THNN_(ReLU6_updateGradInput)(
          THNNState *state,
          THTensor *input,
          THTensor *gradOutput,
          THTensor *gradInput,
          bool inplace)
{
  if (inplace)
  {
    TH_TENSOR_APPLY2(real, gradOutput, real, input,
      if ((*input_data) <= 0 || (*input_data) >= 6)
        *gradOutput_data = 0;
    );
    THTensor_(set)(gradInput, gradOutput);
  }
  else
  {
    THTensor_(resizeAs)(gradInput, input);
    TH_TENSOR_APPLY3(real, gradInput, real, gradOutput, real, input,
      if ((*input_data) > 0 && (*input_data) < 6)
        *gradInput_data = *gradOutput_data;
      else
        *gradInput_data = 0;
    );
  }
}

#endif

File: torch/legacy/nn/ReLU6.py
import torch
from .Module import Module


class ReLU6(Module):

    def __init__(self, inplace=False):
        super(ReLU6, self).__init__()
        self.inplace = inplace

    def updateOutput(self, input):
        self._backend.HardTanh_updateOutput(
            self._backend.library_state,
            input,
            self.output,
            0, 6, self.inplace
        )
        return self.output

    def updateGradInput(self, input, gradOutput):
        self._backend.HardTanh_updateGradInput(
            self._backend.library_state,
            input,
            gradOutput,
            self.gradInput,
            0, 6, self.inplace
        )
        return self.gradInput
